horace_core/utilities/split_data_blocks.m

```diff
@@ -30,74 +32,122 @@
     cumulative_sum = [];
     return;
 end
-validateattributes(start_val, {'numeric'}, {'vector', 'positive'});
+validateattributes(start_pos, {'numeric'}, {'vector', 'positive'});
 validateattributes(block_sizes, {'numeric'}, {'vector', 'positive'});
-validateattributes(max_chunk_sum, {'numeric'}, {'scalar', 'positive'});
-if any(size(start_val)~=size(block_sizes))
+validateattributes(buf_size, {'numeric'}, {'scalar', 'positive'});
+if any(size(start_pos)~=size(block_sizes))
     error('HORACE:utilities:invalid_argument',...
         'The size of start_val array and size of block_sizes array have to be the same');
 end
-if size(start_val,1)>size(start_val,2) % we want it to be rows
-    start_val = start_val';    
+if size(start_pos,1)>size(start_pos,2) % we want it to be rows
+    start_pos = start_pos';
 end
 if size(block_sizes,1)>size(block_sizes,2) % we want it to be rows
-    block_sizes = block_sizes';    
+    block_sizes = block_sizes';
+end
+% glue adjacent data blocks
+cumulative_sum = cumsum(block_sizes);
+last_block_pos = block_sizes+start_pos;
+% remove border between adjacent blocks
+remove_border  = start_pos(2:end)==last_block_pos(1:end-1);
+if any(remove_border)
+    % keep boder positions, which should not be removed
+    start_pos= start_pos([true,~remove_border]);
+    % number of blocks to keep is equal to the number of borders +1
+    % n_unique_blocks = sum(~remove_border)+1;
+    n_blocks = numel(block_sizes);
+    block_ind = zeros(1,n_blocks);
+    cur_ind = 1;
+    block_ind(1) = cur_ind;
+    for i=2:n_blocks
+        if ~remove_border(i-1)
+            cur_ind = cur_ind+1;
+        end
+        block_ind(i) = cur_ind;
+    end
+    block_sizes  = accumarray(block_ind',block_sizes')';
+    cumulative_sum = cumsum(block_sizes);
 end
 
+% split blocks not fitting double buffer into separate blocks
+if any(block_sizes>2*buf_size) % split big ranges into parts to fit buffer.
+    [cell_range,cell_offcet] = arrayfun(@(rg,offs)split_ranges(rg,offs,buf_size),...
+        block_sizes,start_pos,'UniformOutput',false);
+    block_sizes = [cell_range{:}];
+    block_sizes = cell2mat(block_sizes);
+    start_pos   = [cell_offcet{:}];
+    start_pos   = cell2mat(start_pos);
+    cumulative_sum = cumsum(block_sizes);
+end
 
 
-cumulative_sum = cumsum(block_sizes);
-max_num_chunks = floor(cumulative_sum(end)/max_chunk_sum);
-if max_num_chunks*max_chunk_sum<cumulative_sum(end)
+
+
+max_num_chunks = floor(cumulative_sum(end)/buf_size);
```

Comment:
the comments above were very useful - it would be good to comment the next sections of code as well in the same way to make it clear what they do

Reply:
added but usefulness\efforts of such change is highly questionable